### PR TITLE
Major rewrite of UART documentation in response to issues/1754.

### PR DESCRIPTION
**Pull request by michaelstoops**
Like it says, this is a major rewrite of the UART documentation. Nominally, this is because of /https://github.com/raspberrypi/documentation/issues/1754. But also, I found this documentation confusing when I needed it a couple weeks ago, so I gave a shot at fixing the whole thing.

I got rid of the terminology of "primary" and "secondary" devices, which seemed to appear only in this doc, not in Broadcom's docs or others. Instead, I attempted to consistently refer to hardware devices by their canonical hardware names: UART0, etc. When discussing the serial port on the GPIO header, I referred to it as `/dev/serial0`, because that's how it will appear to the reader in Linux.

As you review this, please look first at the structure and clarity for the expected audience. Does this meet the reader's need? My intent was to give:
* a brief introduction to what a UART is in general, and relate it to other things they may know
* then to discuss the UARTs present in Raspberry Pi devices
* then to discuss applications and other how-to instructions

Once we've settled on the structure, I think we need to look at the accuracy of the information. This is a big change and it contains many assertions of fact. I've attempted to validate them against reality. You can see my methodology and data at https://github.com/michaelstoops/documentation/blob/1754-testing/configuration/uart-testing.md. It should be easy to reproduce if you have the appropriate tools.

Lastly, let's look at minor issues of style: formatting and the like. I had to decide which things to 'quote' and which to mark as `inline code`, which I don't think is quite clear in the style guide. I gave a reasonable attempt, but I also edited and re-edited a lot, so I'm not sure they're all consistent in the end.

Your feedback is welcome.

### CODE REVIEW

**Comment by JamesH65** [comment]
@pelwell Since you, IIRC, were responsible for the previous content (maybe not the actual writing?), what are your thoughts on this PR? The UART stuff has always been a bit difficult to understand, a quick read of this and its does seem to be am improvement.

**Comment by pelwell** [comment]
As the author of the original I'm the wrong person to review this.

**Comment by andrum99** [comment]
Hello. I'm the last person to make any significant changes to this page. The issue we seem to have is that at some point in the past, someone decided it would be a good idea to come up with the concept of 'primary' and 'secondary' UARTs. In reality these designations don't really add much to the documentation, and actually just add extra unnecessary complexity, however I kept them as it actually makes some aspects of the UART config slightly easier to explain.

You don't need to explain what a UART is, or some of the other explanatory text you have added, e.g. baud rate. In fact, explanation of non-Pi specific technical details is generally discouraged in this repo.

Taking the proposed changes as a whole, my view is that you have added more complexity in some areas, and removed complexity from others.

**Comment by michaelstoops** [comment]
> As the author of the original I'm the wrong person to review this.

I'd still be glad to hear your point of view. 😊  If I've crossed you by trashing your work, I apologize. I assume that the choices you made were in good faith. Where I chose differently, combining our views should produce something that's better as a whole.

So let's hear it!

**Comment by andrum99** [comment]
See #1850 which addresses #1754.

(This PR does not actually address that issue).

**Comment by michaelstoops** [comment]
> Hello. I'm the last person to make any significant changes to this page. The issue we seem to have is that at some point in the past, someone decided it would be a good idea to come up with the concept of 'primary' and 'secondary' UARTs. In reality these designations don't really add much to the documentation, and actually just add extra unnecessary complexity, however I kept them as it actually makes some aspects of the UART config slightly easier to explain.
> 
> You don't need to explain what a UART is, or some of the other explanatory text you have added, e.g. baud rate. In fact, explanation of non-Pi specific technical details is generally discouraged in this repo.
> 
> Taking the proposed changes as a whole, my view is that you have added more complexity in some areas, and removed complexity from others.

I figure that the casual technologist may benefit from the background info, but I appreciate the point that it's really not configuration information, and this is under the configuration directory. Is there a better place for such content? If not, should there be?

**Comment by michaelstoops** [comment]
> See #1850 which addresses #1754.
> 
> (This PR does not actually address that issue).

Thank you. I had a slightly different understanding of the solution to 1754, which I thought was covered.

It seems that your changes might be lost in the re-structuring, so I'm integrating it into my next revision.

**Comment by andrum99** [comment]
> Is there a better place for such content? If not, should there be?

The general rule with this repo is to not document things that are not Pi-specific, and reasonably succinct. Per the contribution policy at https://www.raspberrypi.org/documentation/CONTRIBUTING.md:

> Note that this documentation is intended to be a short and concise set of helpful resources aimed at the majority of users.

(Also note that while there is some content in the repo that is not Pi-specific, e.g. the Linux tutorial section at documentation/linux, this is historical content which was written before there was any other content on the official website other than the blog and the forums. We now have MagPi articles and the Projects website at projects.raspberrypi.org, among others. The documentation will shortly be getting a major overhaul by Alasdair Allan, who wrote the very excellent docs for the Pico and RP2040).

**Comment by JamesH65** [comment]
Just a small clarification - Alasdair is the technical docs manager, and whilst wrote many parts of the Pico docs, there was a huge contribution from the engineers involved, from the HW team who implemented the RP2040 to the software team who wrote the SDK, examples etc, plus input from a graphics designers in the comms team. And some engineers did both HW design and software, so got to write docs for both! Alasdair managed all of us and provided the overall concepts and took in 'suggestions' from above (i.e. Eben and other directors!). Worked really well, the docs are great.

**Comment by aallan** [comment]
> Alasdair is the technical docs manager, and whilst wrote many parts of the Pico docs, there was a huge contribution from the engineers involved, from the HW team who implemented the RP2040 to the software team who wrote the SDK, examples etc, plus input from a graphics designers in the comms team. And some engineers did both HW design and software, so got to write docs for both! 

Good god. I don't think a single person on the planet has the breadth of knowledge to write documentation all the way from getting started to C, through hardware design, to register level description of silicon. If such a person exists, it certainly isn't me. 

The RP2040 documentation took a huge team effort by most of the hardware and software team begin Pico over the course of an entire year. As @JamesH65 says I wrote some of it, mostly the getting started documentation so that would be written from the perspective of someone that hadn't just spent 4 years building the chip, but I certainly didn't do it all!

File: configuration/uart.md [hunk 1]
```diff
@@ -1,99 +1,100 @@
 # UART configuration
 
-There are two types of UART available on the Raspberry Pi -  [PL011](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0183g/index.html) and mini UART. The PL011 is a capable, broadly 16550-compatible UART, while the mini UART has a reduced feature set.
+A [UART](https://en.wikipedia.org/wiki/Universal_asynchronous_receiver-transmitter) is a serial communication device which uses one or more signal wires and a common ground to transmit and receive data between two devices. When computer people talk about 'serial ports', they usually mean UARTs. People from DOS/Windows background may know these as 'COM ports'.
 
-All UARTs on the Raspberry Pi are 3.3V only - damage will occur if they are connected to 5V systems. An adaptor can be used to connect to 5V systems. Alternatively, low-cost USB to 3.3V serial adaptors are available from various third parties.
+## Hardware
 
-## Pi Zero, 1, 2 and 3 - two UARTs
+On all Raspberry Pi computers except microcontrollers, there are at least two UARTs built into the [SoC](https://en.wikipedia.org/wiki/System_on_a_chip). The Raspberry Pi 4 family has six UARTs.
```
Comment (lurch) [inline configuration/uart.md:7]: The Raspberry Pi Pico is a microcontroller, and so doesn't fall into the "Raspberry Pi computers" category :wink: 
**All** the Pico documentation is at https://www.raspberrypi.org/documentation/rp2040/getting-started/ and the rest of the documentation repo only covers "Raspberry Pi computers" without any mention of the Raspberry Pi microcontroller.

(fun fact: the Raspberry Pi Pico also has 2 UARTs :laughing: )

File: configuration/uart.md [hunk 2]
```diff
@@ -1,99 +1,100 @@
...
 
-The Raspberry Pi Zero, 1, 2, and 3 each contain two UARTs as follows:
+There are also two different types of UART devices available on the Raspberry Pi:
+* [PL011](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0183g/index.html) part of the ARM spec and is patterned on the full-featured [16550 UART](https://en.wikipedia.org/wiki/16550_UART) chip.
```
Comment (lurch) [inline configuration/uart.md:10]: To be pedantic: not "part of the ARM spec", just a peripheral block available from ARM which plugs in nicely to the rest of the AMBA ecosystem.

**Comment by lurch** [comment]
The changes in this PR definitely feel far too wordy - have a look at some of the other documentation pages to get a feel for the typical level of detail used.

File: configuration/uart.md (hunk 1 above)
Comment (aallan) [reply configuration/uart.md:7]: This. There should be no mention of Pico in the mainline documentation. That split happens (will happen) further up the documentation tree.

**Comment by michaelstoops** [comment]
Thank you all for the feedback. I'll give this another edit and come back.

File: configuration/uart.md (hunk 1 above)
Comment (michaelstoops) [reply configuration/uart.md:7]: Great, thanks for the context. This is adjusted in the next rev.

File: configuration/uart.md (hunk 2 above)
Comment (michaelstoops) [reply configuration/uart.md:10]: Thanks, will be changed in next rev.

**Comment by michaelstoops** [comment]
OK, revision made! I took out the background info. I also took out some of the wordy prose and replaced it with brief statements of fact. Hopefully that's a bit closer to where we're trying to go, and also an improvement over the past.

Feedback welcome. :)

File: configuration/uart.md [hunk 3]
```diff
@@ -21,31 +17,33 @@ Broadcom gives each UART a name and number in their SoC documentation, and this
 | BCM2837 | PL011 | mini UART |
 | BCM2711 | PL011 | mini UART | PL011 | PL011 | PL011 | PL011 |
 
- See the [Raspberry Pi 4 Model B Datasheet](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2711/rpi_DATA_2711_1p0_preliminary.pdf) section _5.1.2 GPIO Alternate Functions_, for details on exactly which UART can appear on which GPIO numbers. Earlier models are similar, but have only the first two UARTs. Take care to line up the numbers: TXD0 and RXD0 always belong to UART0, TXD1 to UART1 and so on. Also, remember that the GPIO numbering is on the CPU, and is different than the GPIO pin numbers on the 40-pin J8 header.
+Inside the Raspberry Pi, the UARTs are able to connect in several ways. UART0 and UART1 can connect to either the GPIO lines, or the onboard Bluetooth module, if there is one. UARTs two through five connect to the GPIO lines and nowhere else.
```
Comment (lurch) [inline configuration/uart.md:20]: "GPIO lines" -> "GPIO header" ?

File: configuration/uart.md [hunk 4]
```diff
@@ -21,31 +17,33 @@ Broadcom gives each UART a name and number in their SoC documentation, and this
...
 
-Inside the Raspberry Pi, the UARTs are able to connect in several ways. UART0 and UART1 can connect to either the GPIO lines or the onboard Bluetooth module, if there is one. UARTs two through five connect to the GPIO lines and nowhere else.
+See the [Raspberry Pi 4 Model B Datasheet](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2711/rpi_DATA_2711_1p0_preliminary.pdf) section _5.1.2 GPIO Alternate Functions_, for pinout details. Earlier models are similar but with only the first two UARTs.
```
Comment (lurch) [inline configuration/uart.md:22]: IIRC the earlier models have even more flexible mapping of where the UARTs end up? So maybe it's worth adding a link to the BCM2835 datasheet here?

File: configuration/uart.md [hunk 5]
```diff
@@ -21,31 +17,33 @@ Broadcom gives each UART a name and number in their SoC documentation, and this
...
-
-There are chips called ['level shifters'](https://en.wikipedia.org/wiki/Level_shifter) which convert voltages between levels. You can find them either as individual ICs for using in your own designs, or integrated into other devices to make them compatible. They're inexpensive and readily available from third parties.
+All the onboard Raspberry Pi UARTs are 3.3V ([CMOS](https://en.wikipedia.org/wiki/CMOS) [logic level](https://en.wikipedia.org/wiki/Logic_level)). Do not connect Raspberry Pi UARTs to 5V ([TTL logic level](https://en.wikipedia.org/wiki/Transistor%E2%80%93transistor_logic)) circuits, because the Raspberry Pi will be damaged. [Level shifters](https://en.wikipedia.org/wiki/Level_shifter) are inexpensive and readily available.
```
Comment (lurch) [inline configuration/uart.md:26]: Maybe worth mentioning 3.3V USB UART adaptors too? (as I suspect they'll be a more common approach than using a level-shifter)

Is the distinction between CMOS and TTL (rather than just between 3.3V and 5V) necessary? :shrug: (I'm not a hardware engineer :wink: )

File: configuration/uart.md [hunk 6]
```diff
@@ -21,31 +17,33 @@ Broadcom gives each UART a name and number in their SoC documentation, and this
...
 
 ## Software support
 
-Raspberry Pi OS has built-in drivers for the UARTs built into the SoC, and common external types as well.
+Raspberry Pi OS ships with drivers for the onboard UARTs.
 
-The PL011 devices will appear as [character devices](https://en.wikipedia.org/wiki/Device_file#Character_devices) with names like `/dev/ttyAMA*`. If there are multiple PL011s, they're in the same _order_ as the Broadcom documentation, but not necessarily the same _number_. The mini UART always appears as character device named `/dev/ttyS0`, and there is never more than one. If you want to be specific about which hardware type you're using, use the `/dev/ttyAMA*` and `/dev/ttyS0` names.
+PL011s appear as [character devices](https://en.wikipedia.org/wiki/Device_file#Character_devices) `/dev/ttyAMA*`. If there are multiple PL011s, they're in the same _order_ as the Broadcom documentation, but not necessarily the same _number_. The mini UART always appears as `/dev/ttyS0`. If you want to be specific about hardware types, use the `/dev/ttyAMA*` and `/dev/ttyS0` names.
```
Comment (lurch) [inline configuration/uart.md:32]: I'm not sure the character device link will be useful for most people?  (presumably UARTs are _always_ character devices?)
"The PL011 UARTs appear as `/dev/ttyAMA*`."

File: configuration/uart.md [hunk 7]
```diff
@@ -21,31 +17,33 @@ Broadcom gives each UART a name and number in their SoC documentation, and this
...
 
-There are also be symbolic links with names like `/dev/serial*`. If a UART is enabled and configured to appear on pins 8 and 10 of the GPIO header, `/dev/serial0` will link to that UART. If a UART is enabled and configured to connect to the onboard Bluetooth module, `/dev/serial1` links to that one. If you want to be specific about where a UART is connected, use the `/dev/serial*` names.
+There are also be symbolic links: `/dev/serial*`. If a UART is enabled and connected to pins 8 and 10 of the GPIO header, `/dev/serial0` will link to that UART. If a UART is enabled and connected to the onboard Bluetooth module, `/dev/serial1` links to that one. If you want to be specific about a UART connection, use the `/dev/serial*` names.
```
Comment (lurch) [inline configuration/uart.md:34]: "also be" -> "also"

Perhaps "connected" -> "internally connected"? (to differentiate it from anything a user might externally connect to the GPIO header)

Perhaps "link to" -> "point to"? :shrug: 

I'm not sure if the "If you want to be specific..." sentences (in both this and the previous paragraph) are useful - surely a user has to know where their UART is physically connected to, and so the `/dev/serial*` symlinks should always be preferred?

File: configuration/uart.md [hunk 8]
```diff
@@ -21,31 +17,33 @@ Broadcom gives each UART a name and number in their SoC documentation, and this
...
 
 The Linux device management system chooses which hardware UARTs are connected to the GPIO pins and creates the `/dev/serial*` symlinks. By default, if there is an onboard Bluetooth module, it gets UART0 and the GPIO header gets UART1. If there is no Bluetooth, the GPIO header gets UART0. This behavior is configurable and can be changed with device tree overlays.
```
Comment (lurch) [inline configuration/uart.md:36]: "which hardware UARTs are connected to which GPIO pins" ?

_Perhaps_ worth using "UART1 (mini UART)" and "UART0 (PL011)" ? Or maybe that just adds more confusion.

File: configuration/uart.md [hunk 9]
```diff
@@ -57,7 +55,7 @@ There are several ways to configure the UART hardware.
 
 #### config.txt
 
-Adding `enable-uart=1` to `/boot/config.txt` should be enough to enable the first two UART ports.
+On most devices, adding `enable-uart=1` to `/boot/config.txt` should be enough to enable the first two UART ports. On compute modules, you must also add a device-tree overlay eo enable the particular UART you want to use.
```
Comment (lurch) [inline configuration/uart.md:58]: I'm not sure if "enable the first two UART ports" is actually what this option does? IIRC it's for ensuring the UART on the GPIO header is usable?

File: configuration/uart.md [hunk 10]
```diff
@@ -114,31 +110,16 @@ N.B. Selecting the wrong early console can prevent the Pi from booting.
 
 ### Raspberry Pi as terminal
 
-A device you use to login to a serial console is called a ['terminal'](https://en.wikipedia.org/wiki/Computer_terminal). A Raspberry Pi can also act as a terminal, so you can actually login from one Raspberry Pi to another with just a few wires and no network! You just need to connect the two devices' grounds together, and then connect one's TXD to the other's RXD, and vice versa. No resistors or other special hardware are required.
+A Raspberry Pi can also act as a serial terminal, which can be used to connect to another device's serial console. You'll need to disable the terminal's serial console so that the serial console and terminal program won't conflict with each other.
```
Comment (lurch) [inline configuration/uart.md:113]: "terminal's serial console" -> "kernel's serial console" ?

**Comment by lurch** [comment]
Thanks for your extensive edits @michaelstoops , this "feels" much better now :+1:

File: configuration/uart.md (hunk 3 above)
Comment (michaelstoops) [reply configuration/uart.md:20]: Yes. Will appear on the next rev.

File: configuration/uart.md (hunk 4 above)
Comment (michaelstoops) [reply configuration/uart.md:22]: I think we're looking at pages 102 and 176 of this document: https://www.raspberrypi.org/app/uploads/2012/02/BCM2835-ARM-Peripherals.pdf.

It looks like there are a couple other available GPIO mappings for UART0, but none of them on the GPIO header. I don't think those bear mention.

File: configuration/uart.md (hunk 5 above)
Comment (michaelstoops) [reply configuration/uart.md:26]: I had included the names "TTL" and "CMOS" because product descriptions often express these names rather than the numerical voltage specs. An internet search for "USB serial port" shows quite a few products listed as "TTL" but none as "CMOS". Additionally, I had thought that TTL is pretty explicitly 5V logic levels, but many of the devices listed as TTL claim to be 3.3V and 5V compatible--if I read them correctly. I have to think they signal with 3.3V and hope it's close enough to be correctly interpreted by a 5V interface.

Considering the ambiguity of "TTL" and the absence of "CMOS" terminology, I think we're better off dropping those names and going with the numeric voltages.

This will be corrected in the next rev.

File: configuration/uart.md (hunk 6 above)
Comment (michaelstoops) [reply configuration/uart.md:32]: Indeed, any UART will be a character device, and those in the know already know that. I expressed it that way to draw the connection between the hardware, the kernel and filesystem. While a user will refer to the device by path, and that is the default path, it's the device node properties that connect to the hardware.

I think the link does no harm, and offers a convenient path to knowledge for who don't know what it means. I say leave the link.

File: configuration/uart.md (hunk 7 above)
Comment (michaelstoops) [reply configuration/uart.md:34]: > "also be" -> "also"

Yes

> Perhaps "connected" -> "internally connected"? (to differentiate it from anything a user might externally connect to the GPIO header)

Sure.

> I'm not sure if the "If you want to be specific..." sentences (in both this and the previous paragraph) are useful - surely a user has to know where their UART is physically connected to, and so the /dev/serial* symlinks should always be preferred?

I think finding the right name is really important, and I think you're right that we should point readers to `/dev/serial*`. My intention was to explain the relationship between the names and the hardware. The next commit will include a rewrite which puts `/dev/serial*`.

File: configuration/uart.md (hunk 8 above)
Comment (michaelstoops) [reply configuration/uart.md:36]: Rewritten in next commit.

File: configuration/uart.md (hunk 9 above)
Comment (michaelstoops) [reply configuration/uart.md:58]: I think you're right. Is the source code available to have a look and know for sure?

File: configuration/uart.md (hunk 10 above)
Comment (michaelstoops) [reply configuration/uart.md:113]: I meant "terminal" to refer to the device acting as the terminal, as opposed to the device exposing a serial console. But it was hard to find the right word.

I've revised it in the next commit. Let's see if it's any clearer.

**Comment by michaelstoops** [comment]
I think we're closing in on it. :)

Feedback still welcome, as always.

File: configuration/uart.md (hunk 5 above)
Comment (aallan) [reply configuration/uart.md]: > All the onboard Raspberry Pi UARTs are 3.3V ([CMOS](https://en.wikipedia.org/wiki/CMOS) [logic level](https://en.wikipedia.org/wiki/Logic_level)). Do not connect Raspberry Pi UARTs to 5V ([TTL logic level](https://en.wikipedia.org/wiki/Transistor%E2%80%93transistor_logic)) circuits

The TTL connectors I've dealt with in the past use 3.3V logic, not 5V.

File: configuration/uart.md (hunk 9 above)
Comment (lurch) [reply configuration/uart.md]: `config.txt` is handled by `start.elf`, and that's closed source.

File: configuration/uart.md (hunk 4 above)
Comment (lurch) [reply configuration/uart.md:22]: Ahh you're right, I'd forgotten that those changes were outside of the GPIOs available on the header.

**Comment by michaelstoops** [comment]
@aallan, can I get a signal from you on whether this edit falls in line with where the documentation is going? Is this one headed for a merge?

File: configuration/uart.md [hunk 11]
```diff
@@ -1,99 +1,96 @@
...
-In order to use the mini UART, you need to configure the Raspberry Pi to use a fixed VPU core clock frequency. This is because the mini UART clock is linked to the VPU core clock, so that when the core clock frequency changes, the UART baud rate will also change. The `enable_uart` and `core_freq` settings can be added to `config.txt` to change the behaviour of the mini UART. The following table summarises the possible combinations:
+#### config.txt
 
-| Mini UART set to | core clock | Result |
-|------------------|------------|--------|
-| primary UART     | variable   | mini UART disabled |
-| primary UART     | fixed by setting `enable_uart=1` | mini UART enabled, core clock fixed to 250MHz, or if `force_turbo=1` is set, the VPU turbo frequency |
-| secondary UART   | variable   | mini UART disabled |
-| secondary UART   | fixed by setting `core_freq=250` | mini UART enabled |
+On most devices, adding `enable-uart=1` to `/boot/config.txt` should ensure that a UART is available on the GPIO header. On compute modules, you must also add a device-tree overlay eo enable the particular UART you want to use.
```
Comment (JamesH65) [inline configuration/uart.md:58]: Typo 'eo'

**Comment by JamesH65** [comment]
This reads better than the original and has more information, so its a good change in my book, but I noted a couple of issues that I have commented on inline.

File: configuration/uart.md [hunk 12]
```diff
@@ -111,24 +108,22 @@ The baudrate is set to 115200.
...
 
-For full instructions on how to use Device Tree overlays see [this page](device-tree.md). In brief, add a line to the `config.txt` file to apply a Device Tree overlay. Note that the `-overlay.dts` part of the filename is removed. For example:
-```
-dtoverlay=disable-bt
-```
 
 ## Relevant differences between PL011 and mini UART
 
-The mini UART has smaller FIFOs. Combined with the lack of flow control, this makes it more prone to losing characters at higher baudrates. It is also generally less capable than a PL011, mainly due to its baud rate link to the VPU clock speed.
+The mini UART measures time based on the CPU's core clock. If the core clock frequency is allowed to change, the mini UART's baud rate will change with it, and will not be what was intended.
```
[the remark this answers is not in the document]
Comment (michaelstoops) [reply configuration/uart.md]: Fixed. Thanks for the clarification.

File: configuration/uart.md (hunk 11 above)
Comment (michaelstoops) [reply configuration/uart.md]: Fixed, good catch.

**Comment by michaelstoops** [comment]
> This reads better than the original and has more information, so its a good change in my book, but I noted a couple of issues that I have commented on inline.

Fixed. Thanks!

File: configuration/uart.md [hunk 13]
```diff
@@ -1,99 +1,96 @@
...
+| BCM2835 | PL011 | mini UART |
+| BCM2836 | PL011 | mini UART |
+| BCM2837 | PL011 | mini UART |
+| BCM2711 | PL011 | mini UART | PL011 | PL011 | PL011 | PL011 |
 
-## Pi 4 - six UARTS
+Inside the Raspberry Pi, the UARTs are able to connect in several ways. UART0 and UART1 can connect to either the GPIO header, or the onboard Bluetooth module, if there is one. UARTs two through five can be enabled and disabled, but have no alternate connections.
 
-The Raspberry Pi 4 has four additional PL011s, which are disabled by default. The full list of UARTs on the Pi 4 is:
+See the [Raspberry Pi 4 Model B Datasheet](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2711/rpi_DATA_2711_1p0_preliminary.pdf) section _5.1.2 GPIO Alternate Functions_, for pinout details. Earlier models are similar but with only the first two UARTs.
```
Comment (lurch) [inline configuration/uart.md:22]: Just noticed that this is using an absolute `https://www.raspberrypi.org/documentation/...` URL. Could you update it to be a relative URL instead please?

Comment (michaelstoops) [reply configuration/uart.md]: Done!

**Comment by aallan** [comment]
There's been a lot of back and forth on this one. What still needs doing here?

I'd encourage you to wrap this PR up in the next week or so as we're in the process of transitioning the documentation from the current Markdown-based source format to Asciidoc. At some point soon — probably around the end of June, beginning of July — we will freeze the current documentation repo. After that time contributions and PRs based on the Markdown source will not be accepted, and any PRs that are still open will be closed.

See #1911 for full details.

**Comment by aallan** [comment]
There is an alternative being offered to this PR in #1850. So we merge this, that, or neither. Anyone have any thoughts?

**Comment by aallan** [comment]
Merged #1950 instead.
